design/proposal-logging.md

```diff
@@ -0,0 +1,188 @@
+
+Tasks that can be split to separate merges (in order):
+
+* Switch all collectors to use goroutines to collect (concurrently).
+* Add a channel to all collectors that can be used to send the list of pods.  Create a new function to collect pod logs, that receives on the channel.  Put that in a goroutine. Set a `WaitGroup` that closes the channel when all the collectors (not the new one that grabs pod logs) are completed.
+* Change each collector to use the new channel to send the list of pods from which to collect logs (one PR at a time?).  Unit tests can be added/updated at this point.
+* Deduplicate the logs collection (if it's not already implemented above)
+* Ensure that sbctl no longer needs the `-c`
+
+
+
+Example code demonstrating the design:
+
+```go
+package main
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type collector interface {
+	Collect(chan []string) string
+}
+
+type someCollector struct {
+	pods []string
+}
+
+func (c *someCollector) Collect(ch chan []string) string {
+
+	fmt.Println("running collect on someCollector")
+	time.Sleep(1 * time.Second)
+	ch <- c.pods
+	return "someCollector finished"
+}
+
+type cephCollector struct {
+	cephpods []string
+}
+
+func (c *cephCollector) Collect(ch chan []string) string {
+	fmt.Println("running collect on cephCollector")
+	time.Sleep(2 * time.Second)
+	ch <- c.cephpods
+	return "cephCollector finished"
+}
+
+type podLogCollector struct {
+}
+
+// function that receives the list of pods to collect logs from
+// alter the logs collector to just send a list here, this is a new function
+// to go collect the actual logs.
+func (p *podLogCollector) GetPodLogs(ch chan []string) string {
+	// consolidate all the pods and dedupe
+	for pod := range ch {
+		time.Sleep(1 * time.Second)
+		fmt.Println("podLogCollector collected logs from", pod)
+		// add the list to an index
+		// if the logs haven't already been collected, go fetch them
+	}
+	return "podLogCollector finished"
+}
+
+func main() {
+	podChan := make(chan []string, 2)
+	collectors := []collector{
+		&someCollector{pods: []string{"pod1", "pod2"}},
+		&cephCollector{cephpods: []string{"rook1", "osd1", "mon1"}},
+		&someCollector{pods: []string{"pod3", "pod4"}},
+		&someCollector{pods: []string{"pod5", "pod6"}},
+	}
+
+	var logsWg sync.WaitGroup
+	logsWg.Add(1)
+	// grab a list of pods from which to collect logs via a channel read
+	logs := &podLogCollector{}
+	go func(l *podLogCollector) {
+		fmt.Println("Running GetPodLogs.... ", l.GetPodLogs(podChan))
+		defer logsWg.Done()
+	}(logs)
+
+	// run the collectors, they need to send the pod list which wants logs to the channel
+	var wg sync.WaitGroup
+	for _, coll := range collectors {
+		wg.Add(1)
+		go func(coll collector) {
+			defer wg.Done()
+			fmt.Println(coll.Collect(podChan))
+		}(coll)
+	}
+
+	wg.Wait()
+	close(podChan)
+	logsWg.Wait()
+}
+```
+
+## Limitations
+ 
+## Assumptions
+
+* All collectors can be run concurrently, none depend on other collectors
```

Comment:
Everything makes sense in this proposal to me, but we may have some edge cases here. I know i've seen users in the past rely on the synchronous nature of how collectors run, and require one to run before the other. An example:

```yaml
    - exec:
        name: generate-output
        collectorName: generate-output
        namespace: '{{repl Namespace}}'
        selector:
          - app=some-selector
        command: ["sh"]
        args: ["-c", "cd /some/data/dir && mv `ls -t | head -n 1` latest"]
        timeout: 5s
    - copy:
        name: collect-output
        collectorName: collect-output
        selector:
          - app=some-selector
        namespace: '{{repl Namespace}}'
        containerPath: /some/data/dir/latest
```

I wouldn't doubt there are other use cases out there as well. May need to think about adding `depends-on` similar to what you see in CI pipelines.

Reply:
re-written to address this (among other things).  Thanks for the feedback!

Reply:
From the docs: "The exec collector can be used to run a command in an existing pod and include the output in the collected data."  Surely if this is the case, we shouldn't need to run the copy after the exec?  I mean, if it needs to create a backup or something, rather than copy the file it shouldn't be writing to the pod at all, and sending the result to stdout.